test_cases/stdlib/check_re.py

```diff
@@ -0,0 +1,23 @@
+from __future__ import annotations
+
+import re
+import typing as t
+from typing_extensions import assert_type
+
+
+def check_search(str_pat: re.Pattern[str], bytes_pat: re.Pattern[bytes]) -> None:
+    assert_type(str_pat.search("x"), t.Optional[t.Match[str]])
+    assert_type(bytes_pat.search(b"x"), t.Optional[t.Match[bytes]])
+    assert_type(bytes_pat.search(bytearray(b"x")), t.Optional[t.Match[bytes]])
```

Comment:
This is an unusual case because of the implicit promotion, maybe also try a type that isn't implicitly promoted (e.g., `pickle.PickleBuffer`).

Reply:
_You_ added this test in #7679, I'm just moving it to a different file in this PR ;)

But I'll make the change!